Add options to DevTools.

workgroupengineering opened the pull request:
…ools window is also mininized.

## What does the pull request do?
<!--- Give a bit of background on the PR here, together with links to with related issues etc. -->
When you minimize the window on which DevTools is connected, the DevTools window is not minimized.

## What is the current behavior?
<!--- If the PR is a fix, describe the current incorrect behavior, otherwise delete this section. -->
If minimize the window on which the DevTools is attached, the DevTools window is also mininized.

## What is the updated/expected behavior with this PR?
<!--- Describe how to test the PR. -->


## How was the solution implemented (if it's not obvious)?
<!--- Include any information that might be of use to a reviewer here. -->


## Checklist

- [ ] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

## Breaking changes
<!--- List any breaking changes here. When the PR is merged please add an entry to https://github.com/AvaloniaUI/Avalonia/wiki/Breaking-Changes -->


## Fixed issues
<!--- If the pull request fixes issue(s) list them like this: 
Fixes #123
Fixes #456
-->
Fixes #5008

MarchingCube commented:
We might need to make this configurable perhaps since from my perspective this PR is introducing a bug that prevents DevTools from staying on top of inspected control (and not minimizing together with said window).

donandren commented:
personally i like very much dev tools is on top and minimized together with the monitored window. But it would make sense this is an configurable option.

workgroupengineering committed:
Add KeyGesture implicit convert form Key and (Key,KeyModifiers)

workgroupengineering committed:
Add implicti connvert form (double,doble) to Size

workgroupengineering committed:
Add DevToolsOptions

workgroupengineering committed:
Implement DevToolsOptions

workgroupengineering commented:
Hi, I have implemented DevToolsOptions. I would like to use it in the following way, but I don't want to break the api signature. Do you have any suggestions?

```csharp
            this.AttachDevTools()
                .With(new Avalonia.Diagnostics.DevToolsOptions()
                {
                    Gesture = (Key.F12, KeyModifiers.Control),
                    Size = (1024,768),
                    OnTop = false,
                });
```

workgroupengineering commented on src/Avalonia.Diagnostics/Diagnostics/DevTools.cs:
```diff
@@ -5,20 +5,28 @@
+#nullable enable 
 namespace Avalonia.Diagnostics
 {
     public static class DevTools
```
Why it is public? Can i make internal?

MarchingCube replied:
Do you mean why `DevTools` class is `public`? In this case - well this is an API for the user to attach it, I think there is no other available way.

workgroupengineering replied:
There is DevToolsExtensions

MarchingCube replied:
Oh, I've forgotten about that extension - in this case we might want to make this part internal to reduce API surface (if both provide the same functionality).

workgroupengineering commented:
@MarchingCube do you have a suggestion for this?

> 
> 
> Hi, I have implemented DevToolsOptions. I would like to use it in the following way, but I don't want to break the api signature. Do you have any suggestions?
> 
> ```cs
>             this.AttachDevTools()
>                 .With(new Avalonia.Diagnostics.DevToolsOptions()
>                 {
>                     Gesture = (Key.F12, KeyModifiers.Control),
>                     Size = (1024,768),
>                     OnTop = false,
>                 });
> ```

MarchingCube commented on src/Avalonia.Diagnostics/Diagnostics/DevTools.cs:
```diff
@@ -29,6 +37,11 @@ void PreviewKeyDown(object sender, KeyEventArgs e)
+            return Open(root, new DevToolsOptions());
+        }
+
+        static IDisposable Open(TopLevel root, DevToolsOptions options)
```
Missing access modifier.

MarchingCube reviewed (commented):
I'm generally fine with this PR, not sure about extra implicit operators for `Size` and `KeyGesture` since we had a similar change request at some point and it was declined from what I recall (@grokys - what is your stance on this?).

MarchingCube commented:
> @MarchingCube do you have a suggestion for this?
> 
> > Hi, I have implemented DevToolsOptions. I would like to use it in the following way, but I don't want to break the api signature. Do you have any suggestions?
> > ```cs
> >             this.AttachDevTools()
> >                 .With(new Avalonia.Diagnostics.DevToolsOptions()
> >                 {
> >                     Gesture = (Key.F12, KeyModifiers.Control),
> >                     Size = (1024,768),
> >                     OnTop = false,
> >                 });
> > ```

I think we should avoid breaking signature in this case and I wonder if it is even worth it to have such "builder" interface here (as this is a fairly niche API) especially since it requires breaking changes.

workgroupengineering committed:
Merge branch 'master' into fixes/Issue_5008

workgroupengineering committed:
Changed DevTools access modifier

workgroupengineering committed:
Add AttachDevTools overload for expose DevToolsOptions

workgroupengineering committed:
DevToolsOptions add addtional options

- ShouldVisualizeMarginPadding
- ShouldVisualizeDirtyRects
- ShowFpsOverlay
- ShowConsole
- ShowLayoutVisualizer

workgroupengineering commented:
Hi,
> 
> 
> I'm generally fine with this PR, not sure about extra implicit operators for `Size` and `KeyGesture` since we had a similar change request at some point and it was declined from what I recall (@grokys - what is your stance on this?).

if you want, I strip it from the PR. It is just to simplify the writing.



> I think we should avoid breaking signature in this case and I wonder if it is even worth it to have such "builder" interface here (as this is a fairly niche API) especially since it requires breaking changes.

I too am for not breaking signatures. On the "builder" pattern it would help to extend more easily and limit breaking changes.

I found a workaround but it limits the scope of AttachDevTools in the WindowBase subclass.

```csharp
        public static IDevTools AttachDevTools(this WindowBase root)
        {
            ...
        }
```

workgroupengineering committed:
Merge branch 'master' into fixes/Issue_5008

grokys commented:
@workgroupengineering what's the background on this? What use-case is it fixing? As the others have commented, I find the current behavior the desired behavior and before adding more options I'd like to know what problem it was causing you.

workgroupengineering commented:
Hy @grokys 

sorry if I answer you just now but I have been very busy in another project.

The scenario is this:
development machine with two monitors, on the main visual studio and the application that I am debugging, on the second monitor the DevTools.

Very often I have to do a step by step debug and see the behavior in the DevTools. To debug step by step I am forced to minimize the window but at the same time the DevTools window is minimized.

This PR in addition to adding the option to handle this behavior adds other amounts of options that avoid numerous clicks when starting the DevTools.

The other options are:
| Option | Description |
|--------|-------------|
|Size|The size of DevTools window. Default value is 1024x512.|
| ShowLayoutVisualizer | Indicates whether to display Layout Visualizer, default value is true.. it's useful on small screens when you don't care to see the layout. |
| ShouldVisualizeMarginPadding   | Indicates whether to display margins and padding, default value is true. |
| ShouldVisualizeDirtyRects |  Indicates whether to display dirty rect, default value is false. |
| ShowFpsOverlay |  Indicates whether to display fps overly, default value is false. |
| ShowConsole | Indicates whether to display console, default value is false.|

If you feel that the option that allows you to change is not necessary I remove it, but the other options I think it is good to accept them.

grokys commented on src/Avalonia.Diagnostics/Diagnostics/DevTools.cs:
```diff
@@ -5,20 +5,28 @@
 namespace Avalonia.Diagnostics
 {
-    public static class DevTools
+    internal static class DevTools
```
This is a breaking change, meaning that we couldn't merge this in the 0.10 releases.

grokys commented on src/Avalonia.Diagnostics/Diagnostics/DevTools.cs:
```diff
@@ -28,7 +36,7 @@ void PreviewKeyDown(object sender, KeyEventArgs e)
         }
 
-        public static IDisposable Open(TopLevel root)
+        private static IDisposable Open(TopLevel root, DevToolsOptions options)
```
Again, a breaking change. This method was public to allow showing DevTools immediately without pressing a key.

grokys commented on src/Avalonia.Diagnostics/Diagnostics/DevTools.cs:
```diff
@@ -38,15 +46,21 @@ public static IDisposable Open(TopLevel root)
                 };
 
                 window.Closed += DevToolsClosed;
+                window.ViewModel.ShouldVisualizeDirtyRects = options.ShouldVisualizeDirtyRects;
```
This can already be controlled by using `window.Renderer.VisualizeDirtyRects` - I don't think it's necessary here; DevTools just provides a UI to toggle this property, if you want to control it programmatically you should use the `Renderer` property.

grokys commented on src/Avalonia.Diagnostics/Diagnostics/DevTools.cs:
```diff
@@ -38,15 +46,21 @@ public static IDisposable Open(TopLevel root)
+                window.ViewModel.ShouldVisualizeDirtyRects = options.ShouldVisualizeDirtyRects;
+                window.ViewModel.ShouldVisualizeMarginPadding = options.ShouldVisualizeMarginPadding;
+                window.ViewModel.Console.IsVisible = options.ShowConsole;
+                window.ViewModel.ShowFpsOverlay = options.ShowFpsOverlay;
```
Same with this: `window.Renderer.DrawFps`.

workgroupengineering replied:
I know they are already in the menu, but it's boring to set every FPS or DirtyRects , or whether or not to view the console.

grokys commented on src/Avalonia.Diagnostics/Diagnostics/DevTools.cs:
```diff
@@ -38,15 +46,21 @@ public static IDisposable Open(TopLevel root)
                 s_open.Add(root, window);
 
-                if (root is Window inspectedWindow)
+                if (options.OnTop && root is Window inspectedWindow)
```
Not sure `OnTop` is the correct terminology here: "Topmost" windows have a specific meaning and but this switch isn't controlling that - it's controlling whether DevTools is shown as a child window.

grokys commented on src/Avalonia.Diagnostics/Diagnostics/DevTools.cs:
```diff
@@ -38,15 +46,21 @@ public static IDisposable Open(TopLevel root)
+                window.ViewModel.ShouldVisualizeMarginPadding = options.ShouldVisualizeMarginPadding;
+                window.ViewModel.Console.IsVisible = options.ShowConsole;
+                window.ViewModel.ShowFpsOverlay = options.ShowFpsOverlay;
+                window.ViewModel.ShowLayoutVisualizer = options.ShowLayoutVisualizer;                
```
The side panel now also shows a style browser so the name of this needs to be changed.

workgroupengineering replied:
On small screens LayoutVisualizer and StyleVisualizer making the datagrid with the properties unusable because the available space is small.

grokys replied:
Yeah we know we need to fix the layout of DevTools. But that's a separate PR I think.

grokys commented on src/Avalonia.Diagnostics/Diagnostics/ViewModels/MainViewModel.cs:
```diff
@@ -185,5 +186,9 @@ private void KeyboardPropertyChanged(object sender, PropertyChangedEventArgs e)
+
+        public bool ShowLayoutVisualizer { get => _showLayoutVisualizzer; set => RaiseAndSetIfChanged(ref _showLayoutVisualizzer, value); }
+
+        void ToggleShowLayoutVisualizer(object paramter) => ShowLayoutVisualizer = !ShowLayoutVisualizer;
```
This method doesn't seem to be used?

grokys commented on src/Avalonia.Input/KeyGesture.cs:
```diff
@@ -61,6 +61,16 @@ public override int GetHashCode()
             return !Equals(left, right);
         }
 
+        static public implicit operator KeyGesture ((Key key, KeyModifiers modifiers) value)
```
Adding a implicit operators is a [pretty big deal](https://stackoverflow.com/questions/1782503/implicit-operator-when-is-it-a-good-bad-idea) and should be discussed separately.

grokys commented on src/Avalonia.Visuals/Size.cs:
```diff
@@ -61,6 +61,15 @@ public Size(double width, double height)
+        ///  Implicit convert from tuple of two double (double,double) to <see cref="Size"/>.
+        /// </summary>
+        /// <param name="value">tuple of two <see cref="double"/></param>
+        public static implicit operator Size((double width, double height) value)
```
Adding implicit operators to/from tuples was already discussed in https://github.com/AvaloniaUI/Avalonia/pull/4589 and we decided to not implement for the moment. If we wanted to revisit that decision it should be done in an issue/PR not in an unrelated PR.

grokys commented:
I like the `DevToolsOptions` idea @workgroupengineering but the implementation here needs some work. Given that I don't have much time right now to do reviews it may be quicker for me to just fix up this PR myself. Would that be acceptable?

grokys committed:
Merge branch 'master' into pr/5011-fixes-issue-5008

grokys commented:
I was unable to push my changes to your branch so opened a new PR at #5669.
